ngrok/src/config/policies.rs

```diff
@@ -0,0 +1,250 @@
+    inbound: Vec<Policy>,
+    outbound: Vec<Policy>,
+}
+
+/// A policy that defines rules that should be applied
+#[derive(Serialize, Deserialize, Debug, Clone, Default, PartialEq)]
+#[serde(default)]
+pub struct Policy {
+    name: String,
+    expressions: Vec<String>,
+    actions: Vec<Action>,
+}
+
+/// An action that should be taken if the policy matches
+#[derive(Serialize, Deserialize, Debug, Clone, Default, PartialEq)]
+#[serde(default)]
+pub struct Action {
+    #[serde(rename = "type")]
+    type_: String,
+    config: Option<serde_json::Value>,
+}
+
+/// Error representing invalid string for Policies
+#[derive(Debug, Clone, Error)]
+#[error("invalid policies, err: {}", .0)]
+pub struct InvalidPolicies(String);
+
+/// Used just for json parsing
+#[derive(Serialize, Deserialize, Debug, Clone, Default, PartialEq)]
+struct Envelope {
+    policies: Policies,
+}
+
+impl Policies {
+    /// Create a new empty [Policies] struct
+    pub fn new() -> Self {
+        Policies {
+            ..Default::default()
+        }
+    }
+
+    /// Create a new [Policies] from a json string
+    pub fn from_json(json: impl Into<String>) -> Result<Self, InvalidPolicies> {
+        let envelope: Envelope =
+            serde_json::from_str(&json.into()).map_err(|e| InvalidPolicies(e.to_string()))?;
+        Ok(envelope.policies)
+    }
+
+    /// Create a new [Policies] from a json file
+    pub fn from_file(json_file_path: impl Into<String>) -> Result<Self, InvalidPolicies> {
+        Policies::from_json(
+            read_to_string(json_file_path.into()).map_err(|e| InvalidPolicies(e.to_string()))?,
+        )
+    }
+
+    /// Convert [Policies] to json string
+    pub fn to_json(&self) -> Result<String, InvalidPolicies> {
+        let envelope = Envelope {
+            policies: self.clone(),
+        };
+        serde_json::to_string(&envelope).map_err(|e| InvalidPolicies(e.to_string()))
+    }
+
+    /// Add an inbound policy
+    pub fn add_inbound(&mut self, policy: impl Borrow<Policy>) -> &mut Self {
+        self.inbound.push(policy.borrow().to_owned());
+        self
+    }
+
+    /// Add an outbound policy
+    pub fn add_outbound(&mut self, policy: impl Borrow<Policy>) -> &mut Self {
+        self.outbound.push(policy.borrow().to_owned());
+        self
+    }
+}
+
+impl Policy {
+    /// Create a new [Policy]
+    pub fn new(name: impl Into<String>) -> Self {
+        Policy {
+            name: name.into(),
+            ..Default::default()
+        }
+    }
+
+    /// Add an expression
+    pub fn add_expression(&mut self, expression: impl Into<String>) -> &mut Self {
+        self.expressions.push(expression.into());
+        self
+    }
+
+    /// Add an action
+    pub fn add_action(&mut self, action: Action) -> &mut Self {
+        self.actions.push(action);
+        self
+    }
+}
+
+impl Action {
+    /// Create a new [Action]
+    pub fn new(
+        type_: impl Into<String>,
+        config: Option<impl Into<String>>,
```

Comment:
Also, should we even bother parsing this, or just let the backend parse it and return any errors?

Reply:
I think we need to make sure it's at least valid json, so we can vend it back in `to_json`, since it's raw json objects and not escaped json in a string field.